src/fromager/context.py

```diff
@@ -123,9 +124,39 @@ def add_to_build_order(
             # Requirement and Version instances that can't be
             # converted to JSON without help.
             json.dump(self._build_stack, f, indent=2, default=str)
-        constraints.write_from_build_order(
-            self._constraints_filename, self._build_stack
+        # constraints.write_from_build_order(
+        #     self._constraints_filename, self._build_stack
+        # )
+
+    def update_dependency_graph(
+        self,
+        parent_req: Requirement | None,
+        req_type: str,
+        req: Requirement,
+        version: Version,
+    ):
+        logger.debug(
+            "recording %s dependency %s -> %s %s",
+            req_type,
+            parent_req.name if parent_req else "(toplevel)",
+            req.name,
+            version,
         )
+        # If we have no parent requirement, we are processing an installation
+        # requirement even though it is called "toplevel" elsewhere. So, change
+        # the type. Also set the parent_ßkey to an empty string as a unique value
+        # that can't be mistaken for a real package.
+        if parent_req is None:
+            parent_key = ""
+            req_type = "install"
+        else:
+            parent_key = canonicalize_name(parent_req.name)
+        parent_edges = self.all_edges.setdefault(parent_key, [])
+        parent_edges.append((req_type, canonicalize_name(req.name), str(version)))
+
+        graph_filename = self.work_dir / "graph.json"
+        with open(graph_filename, "w") as f:
+            json.dump(self.all_edges, f, indent=2)
```

Comment:
instead of writing to file each time we want to update the graph, can we do it seperately?

Reply:
We do the same for the build order. The idea is to keep track of what we've produced so far, so it can be analyzed if someone has to debug a build. Given the sorts of things involved in the build, writing the file isn't all that expensive.

Reply:
make sense. maybe instead of opening it and closing it every time, is opening it once globally better? anyways as you said it is a micro-optimization compared to the things we run in the build